src/rpc/blockchain.cpp

```diff
@@ -283,7 +284,22 @@ static RPCHelpMan waitfornewblock()
     NodeContext& node = EnsureAnyNodeContext(request.context);
     Mining& miner = EnsureMining(node);
 
+   // Use tip as reference block, unless a block hash was provided.
     auto block{CHECK_NONFATAL(miner.getTip()).value()};
+    if (!request.params[1].isNull()) {
+        uint256 hash(ParseHashV(request.params[1], "blockhash"));
+        {
+            ChainstateManager& chainman = EnsureAnyChainman(request.context);
+            LOCK(cs_main);
+            const CBlockIndex* block_index{chainman.m_blockman.LookupBlockIndex(hash)};
+            if (!block_index) {
+                throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Block not found");
+            }
+            block.hash = block_index->GetBlockHash();
+            block.height = block_index->nHeight;
+        }
+    }
+
     if (IsRPCRunning()) {
         block = timeout ? miner.waitTipChanged(block.hash, std::chrono::milliseconds(timeout)) : miner.waitTipChanged(block.hash);
     }
```

Comment:
I think it would be good to simplify all of this to just:

```c++
    // Use tip as reference block, unless a block hash was provided.
    uint256 current_tip{request.params[1].isNull()
        ? miner.getTip().value_or(BlockRef{}).hash
        : ParseHashV(request.params[1], "blockhash")};

    BlockRef block{timeout > 0
        ? miner.waitTipChanged(current_tip, std::chrono::milliseconds(timeout))
        : miner.waitTipChanged(current_tip)};
```

In addition to being simpler, this would fix a number of problems and inconveniences of the current implementation:

- If an unrecognized block hash was passed, it would make the RPC return the current block tip instead of throwing an exception.
- If no block hash was passed and tip was null, it would make the RPC wait for a block to be connected instead of throwing an exception.
- It would fix the bug marco described https://github.com/bitcoin/bitcoin/pull/30635#discussion_r1778228488 where this function does not wait in the GUI console.